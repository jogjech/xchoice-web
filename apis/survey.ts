import axios from "axios";
import {
  Survey,
  SurveyResponse,
  SurveyMetadata,
  SurveyReport,
} from "../models/survey";

// For get requests, we can use SWR https://swr.vercel.app/getting-started

interface PostSurveyResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyId?: string;
}

interface PostSurveyResponseResult {
  error?: {
    message: string;
  };
  isError: boolean;
  slug?: string; // slug is the short string which can be used to view / edit this surey respnose
}

interface GetSurveyResult {
  error?: {
    message: string;
  };
  isError: boolean;
  survey?: Survey;
}

interface GetSurveyReportResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyReport?: SurveyReport;
}

interface GetSurveyResponseResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyResponse?: SurveyResponse;
}

interface FindSurveysResult {
  error?: {
    message: string;
  };
  isError: boolean;
  surveyMetadataList?: SurveyMetadata[];
}

const postSurvey = async (
  survey: Survey,
  accessToken: string
): Promise<PostSurveyResult> => {
  console.log("Posting survey");

  const request = {
    title: survey.surveyTitle,
    questions: survey.questions.map((question) => {
      return {
        title: question.questionTitle,
        choices: question.choices.map((choice) => {
          return {
            text: choice.text,
          };
        }),
      };
    }),
  };

  console.log(request);

  try {
    const { data } = await axios.post(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys`,
      request,
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      }
    );

    return {
      isError: false,
      surveyId: data.id,
    };
  } catch (error) {
    return {
      isError: true,
      error: {
        message: error.response.data.message,
      },
      surveyId: "1",
    };
  }
};

const postSurveyResponse = async (
  surveyResponse: SurveyResponse
): Promise<PostSurveyResponseResult> => {
  console.log("Posting survey response", surveyResponse);

  try {
    const { data } = await axios.post(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys/${surveyResponse.surveyId}/responses`,
      surveyResponse.selections
    );

    console.log(data);

    return {
      isError: false,
      slug: data,
    };
  } catch (error) {
    return {
      isError: true,
      error: {
        message: error.response.data.message,
      },
    };
  }
};

const getSurveyResponse = async (
  slug: string
): Promise<GetSurveyResponseResult> => {
  console.log("Getting survey response", slug);
  try {
    const {
      data,
    } = await axios.get(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys/responses`,
      { params: { slug: slug } }
    );

    console.log(data);

    return {
      isError: false,
      surveyResponse: {
        surveyId: data.surveyId,
        selections: data.selections,
      },
    };
  } catch (error) {
    return {
      isError: true,
      error: {
        message: error.response.data.message,
      },
    };
  }
};

const getSurvey = async (
  surveyId: string,
  accessToken?: string
): Promise<GetSurveyResult> => {
  console.log("Getting survey", surveyId);

  try {
    const { data } = await axios.get(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys/${surveyId}`,
      {
        ...(!!accessToken && {
          headers: {
            Authorization: `Bearer ${accessToken}`,
          },
        }),
      }
    );

    console.log(data);

    return {
      isError: false,
      survey: {
        published: true,
        surveyTitle: data.title,
        questions: data.questions.map((question) => {
          return {
            questionTitle: question.title,
            choices: question.choices.map((choice) => {
              return {
                text: choice.text,
              };
            }),
          };
        }),
      },
    };
  } catch (error) {
    return {
      isError: true,
      error: {
        message: error.response.data.message,
      },
    };
  }
};

const getSurveyReport = async (
  surveyId: string,
  accessToken: string
): Promise<GetSurveyReportResult> => {
  console.log(`Getting survey report for survey ${surveyId}`);

  try {
    const { data } = await axios.get(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys/${surveyId}`,
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      }
    );

    console.log(data);

    return {
      isError: false,
      surveyReport: {
        published: true,
        surveyId: surveyId,
        responses: data.totalResponses,
        surveyTitle: data.title,
        questions: data.questions.map((question) => {
          return {
            questionTitle: question.title,
            choices: question.choices.map((choice) => {
              return {
                text: choice.text,
                selections: choice.responses.length,
              };
            }),
          };
        }),
      },
    };
  } catch (error) {
    return {
      isError: true,
      error: {
        message: error.response.data.message,
      },
    };
  }
};

const deleteSurvey = async (surveyId: string) => {};

const findSurveys = async (accessToken: string): Promise<FindSurveysResult> => {
  console.log(`Finding surveys`);

  try {
    const { data } = await axios.get(
      `${process.env.NEXT_PUBLIC_X_CHOICE_API}/surveys`,
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      }
    );

    console.log(data);

    return {
      isError: false,
      surveyMetadataList: data.map((item) => {
        return {
          surveyId: item.surveyId,
          responses: item.responses,
          published: item.published,
          title: item.title,
        };
      }),
    };
  } catch (error) {
    console.log(error);
    return {
      isError: true,
      error: {
        message: "Cannot find surveys for this user",
      },
    };
  }
};

export {
  postSurvey,
  getSurvey,
  getSurveyReport,
  deleteSurvey,
  findSurveys,
  postSurveyResponse,
  getSurveyResponse,
};
