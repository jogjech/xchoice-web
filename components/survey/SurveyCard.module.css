.surveyCard {
  width: 20rem;
  height: 14rem;
  font-size: 1.5rem;
  margin-bottom: 2rem;
}
