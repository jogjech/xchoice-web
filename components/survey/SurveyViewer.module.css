.spinContainer {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.surveyTitle {
  font-size: 2rem;
  margin-bottom: 1rem;
}

.submitButton {
  margin: 1rem 0;
}
