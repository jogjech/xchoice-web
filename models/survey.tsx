export interface Choice {
  text: string;
}

export interface Question {
  questionTitle: string;
  choices: Choice[];
}

export interface Survey {
  surveyTitle: string;
  questions: Question[];
}

export interface SurveyResponse {
  surveyId: string;
  selections: number[];
}

export interface SurveyMetadata {
  surveyId: string;
  responses: number;
  published: boolean;
  userId: string;
  title: string;
}
