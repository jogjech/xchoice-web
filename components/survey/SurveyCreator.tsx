import React, { FunctionComponent, useState } from "react";
import { Form, Input, Button, Result, Alert } from "antd";
import QuestionCreator from "./question/QuestionCreator";
import { postSurvey } from "../../apis/survey";
import { useAuth0 } from "@auth0/auth0-react";
import { useRouter } from "next/router";

interface Props {}

const SurveyCreator: FunctionComponent<Props> = () => {
  const { getAccessTokenSilently } = useAuth0();

  const [form] = Form.useForm();
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const [surveyId, setSurveyId] = useState("");
  const [redirecting, setRedirecting] = useState(false);
  const router = useRouter();

  const handleFormFinish = async (formData: {
    surveyTitle: string;
    questions: [
      {
        questionTitle: string;
        choices: [{ text: string }];
      }
    ];
  }) => {
    console.log(formData);
    setPosting(true);
    setPostError(undefined);

    const accessToken = await getAccessTokenSilently();
    const result = await postSurvey(formData, accessToken);

    setPostError(result.error);
    setPosted(!result.isError);
    setSurveyId(result.surveyId);
    setPosting(false);
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title="Your survey is posted!"
          subTitle={`Survey ID: ${surveyId}.`}
          extra={[
            <Button
              type="primary"
              key="console"
              loading={redirecting}
              onClick={() => {
                setRedirecting(true);
                router.push(`/survey/${surveyId}`);
              }}
            >
              View my survey
            </Button>,
          ]}
        />
      ) : (
        <>
          <div>
            <p className="title">Create your survey</p>
          </div>
          <Form
            layout="vertical"
            form={form}
            initialValues={{
              questions: [
                {
                  choices: [
                    {
                      text: "",
                    },
                  ],
                },
              ],
            }}
            onFinish={handleFormFinish}
            autoComplete="off"
          >
            <p className="sub-title">Basic Info</p>
            <Form.Item
              label="Survey Title"
              name="surveyTitle"
              rules={[{ required: true, message: "Missing survey title" }]}
            >
              <Input placeholder="Survey Title"></Input>
            </Form.Item>
            <QuestionCreator></QuestionCreator>
            <Form.Item>
              <Button type="primary" htmlType="submit" loading={posting}>
                Submit
              </Button>
            </Form.Item>
          </Form>
          {postError && (
            <Alert
              message="Failed to create survey. Please try again."
              type="error"
            ></Alert>
          )}
        </>
      )}
    </>
  );
};

export default SurveyCreator;
