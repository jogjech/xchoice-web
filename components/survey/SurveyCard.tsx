import React, { FunctionComponent, useState } from "react";
import { Card, Statistic, Tag } from "antd";
import { useRouter } from "next/router";
import { SurveyMetadata } from "../../models/survey";
import { FileTextOutlined } from "@ant-design/icons";
import styles from "./SurveyCard.module.css";

interface Props {
  surveyMetadata: SurveyMetadata;
}

const SurveyCard: FunctionComponent<Props> = ({ surveyMetadata }) => {
  const router = useRouter();

  return (
    <>
      {!!surveyMetadata ? (
        <div>
          <Card
            title={surveyMetadata.title}
            hoverable
            bordered
            className={styles.surveyCard}
            headStyle={{
              fontSize: "1.4rem",
              backgroundColor: "rgba(26, 119, 213, 0.6)",
            }}
            onClick={() => {
              router.push(`/survey/${surveyMetadata.surveyId}`);
            }}
          >
            <div>
              <Statistic
                title="Responses"
                value={surveyMetadata.responses}
                prefix={<FileTextOutlined />}
              />
              {surveyMetadata.published ? (
                <Tag color="green">published</Tag>
              ) : (
                <Tag color="orange">unpublished</Tag>
              )}
            </div>
          </Card>
        </div>
      ) : (
        <div>
          <Card loading bordered className={styles.surveyCard}>
            <div></div>
          </Card>
        </div>
      )}
    </>
  );
};

export default SurveyCard;
