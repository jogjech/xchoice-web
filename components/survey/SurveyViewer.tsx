import React, { FunctionComponent, useState, useEffect } from "react";
import { Spin, Button, Result, Alert } from "antd";
import {
  getSurvey,
  postSurveyResponse,
  getSurveyResponse,
} from "../../apis/survey";
import { Survey } from "../../models/survey";
import QuestionViewer from "./question/QuestionViewer";
import { useRouter } from "next/router";
import styles from "./SurveyViewer.module.css";

interface Props {
  surveyId?: string;
  slug?: string;
}

const SurveyViewer: FunctionComponent<Props> = ({
  surveyId: inputSurveyId,
  slug: inputSlug,
}) => {
  // TODO: missing the handle of getSurvey and getSurveyResponse errors
  const [surveyId, setSurveyId] = useState<string>(undefined);
  const [survey, setSurvey] = useState<Survey>(undefined);
  const [loading, setLoading] = useState(true);
  const [validated, setValidated] = useState(false);
  const [selections, setSelections] = useState<number[]>([]);
  const [responseSlug, setResponseSlug] = useState<string>(undefined);
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const router = useRouter();

  useEffect(() => {
    const fetchSurvey = async () => {
      let tempSurveyId: string, tempSelections: number[];

      if (!!inputSurveyId) {
        tempSurveyId = inputSurveyId;
      } else {
        const result = await getSurveyResponse(inputSlug);
        tempSurveyId = result.surveyResponse.surveyId;
        tempSelections = result.surveyResponse.selections;
      }

      setSurveyId(tempSurveyId);

      const result = await getSurvey(tempSurveyId as string);

      setSelections(
        !!tempSelections
          ? tempSelections
          : Array.apply(null, Array(result.survey.questions.length))
      );

      setSurvey(result.survey);
      setLoading(false);
    };
    fetchSurvey();
  }, []);

  const handleSelectionChange = (questionIndex: number) => (
    newValue: number
  ) => {
    console.log(`Question ${questionIndex} received new value ${newValue}`);
    setValidated(false);
    setSelections((prevSelections) => {
      const newSelections = [...prevSelections];
      newSelections[questionIndex] = newValue;
      return newSelections;
    });
  };

  const handleSubmit = async () => {
    console.log("Submitting", selections);
    setPosting(true);
    setPostError(undefined);

    const validationPassed = validateSelections();
    if (validationPassed) {
      console.log("Validation passed");
      const result = await postSurveyResponse({
        surveyId: surveyId,
        selections: selections,
      });
      if (result.isError) {
        setPostError(result.error);
      } else {
        setResponseSlug(result.slug);
      }
      setPosted(!result.isError);
    } else {
      console.log("Validation failed");
    }
    setPosting(false);
  };

  const validateSelections: () => boolean = () => {
    setValidated(true);
    return selections.every((i) => i != null);
  };

  const handleViewResponseButtonClick = () => {
    const toPath = `/survey/response/${responseSlug}`;
    console.log(router);
    if (toPath === router.asPath) {
      router.reload();
    } else {
      router.push(`/survey/response/${responseSlug}`);
    }
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title={`Your response is recorded! Response id: ${responseSlug}`}
          subTitle={`Please use the response id or click the button below to view / edit your response.`}
          extra={[
            <Button
              type="primary"
              key="console"
              onClick={handleViewResponseButtonClick}
            >
              View my response
            </Button>,
          ]}
        />
      ) : loading ? (
        <div className={styles.spinContainer}>
          <Spin tip="Loading survey..."></Spin>
        </div>
      ) : (
        <div>
          <p className={styles.surveyTitle}>{survey.surveyTitle}</p>
          {survey.questions.map((question, index) => (
            <QuestionViewer
              question={question}
              key={index}
              selection={selections[index]}
              handleSelectionChange={handleSelectionChange(index)}
              submitWithNoSelection={validated && selections[index] == null}
            ></QuestionViewer>
          ))}
          <Button
            className={styles.submitButton}
            type="primary"
            onClick={handleSubmit}
            loading={posting}
          >
            Submit
          </Button>
          {postError && (
            <Alert
              message="Failed to post response. Please try again."
              type="error"
            ></Alert>
          )}
        </div>
      )}
    </>
  );
};

export default SurveyViewer;
