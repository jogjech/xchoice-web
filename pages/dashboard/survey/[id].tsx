import React, { FunctionComponent, useState, useEffect } from "react";
import Layout from "../../../components/layout/Layout";
import { Tabs, Breadcrumb, BackTop, Row, Col, Button, Alert } from "antd";
import { Survey } from "../../../models/survey";
import { getSurvey } from "../../../apis/survey";
import SurveyViewer from "../../../components/survey/SurveyViewer";
import ReportViewer from "../../../components/survey/report/ReportViewer";
import { useRouter } from "next/router";

const { TabPane } = Tabs;

interface Props {}

const ManageSurveyPage: FunctionComponent<Props> = () => {
  const router = useRouter();
  const surveyId = router.query.id as string;

  const [loading, setLoading] = useState(true);
  const [survey, setSurvey] = useState<Survey>(undefined);
  const [errorMessage, setErrorMessage] = useState<string>(undefined);

  useEffect(() => {
    if (router.asPath !== router.route) {
      const fetchSurvey = async () => {
        const result = await getSurvey(surveyId);

        if (result.isError) {
          setErrorMessage(result.error.message);
        } else {
          setSurvey(result.survey);
        }

        setLoading(false);
      };
      fetchSurvey();
    }
  }, [router]);

  return (
    <>
      <Layout title="Manage Survey">
        {loading ? (
          <div></div>
        ) : !!errorMessage ? (
          <Alert
            message="Failed to load page. Please refresh to try again."
            type="error"
          ></Alert>
        ) : (
          survey && (
            <>
              <Breadcrumb>
                <Breadcrumb.Item>
                  <a href="/">Home</a>
                </Breadcrumb.Item>
                <Breadcrumb.Item>
                  <a href="/dashboard">Dashboard</a>
                </Breadcrumb.Item>
                <Breadcrumb.Item>{survey.surveyTitle}</Breadcrumb.Item>
              </Breadcrumb>
              <Row justify="space-between" align="bottom">
                <Col>
                  <p className="secondary-title">{survey.surveyTitle}</p>
                </Col>
                <Col style={{ float: "right" }}>
                  {survey.published ? (
                    <Button danger>Unpublish</Button>
                  ) : (
                    <Button type="primary">Publish</Button>
                  )}
                </Col>
              </Row>
              <Tabs
                defaultActiveKey="1"
                centered
                style={{ paddingTop: "1.5rem" }}
              >
                <TabPane tab="Questions" key="1">
                  <BackTop></BackTop>
                  <SurveyViewer surveyId={surveyId} preview></SurveyViewer>
                </TabPane>
                <TabPane tab="Responses Report" key="2">
                  <BackTop></BackTop>
                  <ReportViewer surveyId={surveyId}></ReportViewer>
                </TabPane>
              </Tabs>
            </>
          )
        )}
      </Layout>
    </>
  );
};

export default ManageSurveyPage;
