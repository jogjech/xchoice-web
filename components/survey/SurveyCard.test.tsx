import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { SurveyMetadata } from "../../models/survey";
import SurveyCard from "./SurveyCard";

describe("Survey card", () => {
  let wrapperWithNoMetadata: ShallowWrapper;
  let wrapperWithPublishedMetadata: ShallowWrapper;
  let wrapperWithUnpublishedMetadata: ShallowWrapper;

  const publishedMetadata: SurveyMetadata = {
    surveyId: "surveyId1",
    responses: 13,
    published: true,
    title: "Survey Title 1",
  };

  const unpublishedMetadata: SurveyMetadata = {
    surveyId: "surveyId2",
    responses: 15,
    published: false,
    title: "Survey Title 2",
  };

  beforeEach(() => {
    wrapperWithNoMetadata = shallow(
      <SurveyCard surveyMetadata={null}></SurveyCard>
    );
    wrapperWithPublishedMetadata = shallow(
      <SurveyCard surveyMetadata={publishedMetadata}></SurveyCard>
    );
    wrapperWithUnpublishedMetadata = shallow(
      <SurveyCard surveyMetadata={unpublishedMetadata}></SurveyCard>
    );
  });

  it("with no survey metatdata renders correctly", () => {
    expect(wrapperWithNoMetadata).toMatchSnapshot();
  });

  it("with no survey metatdata renders loading card", () => {
    const card = wrapperWithNoMetadata.find("Card");
    expect(card.prop("loading")).toBeTruthy();
  });

  it("with published survey metatdata renders correctly", () => {
    expect(wrapperWithPublishedMetadata).toMatchSnapshot();
  });

  it("with published survey metatdata renders surey metadata", () => {
    const card = wrapperWithPublishedMetadata.find("Card");
    expect(card.prop("title")).toBe(publishedMetadata.title);

    const statistic = wrapperWithPublishedMetadata.find(
      "withConfigConsumer(Statistic)"
    );
    expect(statistic.prop("title")).toBe("Responses");
    expect(statistic.prop("value")).toBe(publishedMetadata.responses);

    const tag = wrapperWithPublishedMetadata.find("Tag");
    expect(tag.prop("color")).toBe("green");
    expect(tag.text()).toBe("published");
  });

  it("with unpublished survey metatdata renders correctly", () => {
    expect(wrapperWithUnpublishedMetadata).toMatchSnapshot();
  });

  it("with unpublished survey metatdata renders surey metadata", () => {
    const card = wrapperWithUnpublishedMetadata.find("Card");
    expect(card.prop("title")).toBe(unpublishedMetadata.title);

    const statistic = wrapperWithUnpublishedMetadata.find(
      "withConfigConsumer(Statistic)"
    );
    expect(statistic.prop("title")).toBe("Responses");
    expect(statistic.prop("value")).toBe(unpublishedMetadata.responses);

    const tag = wrapperWithUnpublishedMetadata.find("Tag");
    expect(tag.prop("color")).toBe("orange");
    expect(tag.text()).toBe("unpublished");
  });
});
