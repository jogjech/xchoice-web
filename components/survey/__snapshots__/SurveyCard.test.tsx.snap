// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey card with no survey metatdata renders correctly 1`] = `
<Fragment>
  <div>
    <Card
      bordered={true}
      className="surveyCard"
      loading={true}
    >
      <div />
    </Card>
  </div>
</Fragment>
`;

exports[`Survey card with published survey metatdata renders correctly 1`] = `
<Fragment>
  <div>
    <Card
      bordered={true}
      className="surveyCard"
      headStyle={
        Object {
          "backgroundColor": "rgba(26, 119, 213, 0.6)",
          "fontSize": "1.3rem",
        }
      }
      hoverable={true}
      onClick={[Function]}
      title="Survey Title 1"
    >
      <div>
        <withConfigConsumer(Statistic)
          prefix={<ForwardRef(FileTextOutlined) />}
          title="Responses"
          value={13}
        />
        <Tag
          color="green"
        >
          published
        </Tag>
      </div>
    </Card>
  </div>
</Fragment>
`;

exports[`Survey card with unpublished survey metatdata renders correctly 1`] = `
<Fragment>
  <div>
    <Card
      bordered={true}
      className="surveyCard"
      headStyle={
        Object {
          "backgroundColor": "rgba(26, 119, 213, 0.6)",
          "fontSize": "1.3rem",
        }
      }
      hoverable={true}
      onClick={[Function]}
      title="Survey Title 2"
    >
      <div>
        <withConfigConsumer(Statistic)
          prefix={<ForwardRef(FileTextOutlined) />}
          title="Responses"
          value={15}
        />
        <Tag
          color="orange"
        >
          unpublished
        </Tag>
      </div>
    </Card>
  </div>
</Fragment>
`;
